File: deck/base/code/native/kotlin/runtime/cipher.kt
// AES-256-GCM over javax.crypto (built into the JDK, no extra dependency). Key, nonce, plaintext, and ciphertext are
// raw bytes (ByteArray). doFinal on encrypt appends the 128-bit (16-byte) authentication tag, so the output is
// ciphertext || tag, matching the other platforms. Fully qualified (no top-level imports) so the file can be prepended
// as a runtime prelude. Reached only through the public cipher API.
object cipher {
    fun encrypt(key: ByteArray, nonce: ByteArray, plain: ByteArray): ByteArray {
        val instance = javax.crypto.Cipher.getInstance("AES/GCM/NoPadding")
        instance.init(
            javax.crypto.Cipher.ENCRYPT_MODE,
            javax.crypto.spec.SecretKeySpec(key, "AES"),
            javax.crypto.spec.GCMParameterSpec(128, nonce),
        )
        return instance.doFinal(plain)
    }
    fun decrypt(key: ByteArray, nonce: ByteArray, sealed: ByteArray): ByteArray {
        val instance = javax.crypto.Cipher.getInstance("AES/GCM/NoPadding")
        instance.init(
            javax.crypto.Cipher.DECRYPT_MODE,
            javax.crypto.spec.SecretKeySpec(key, "AES"),
            javax.crypto.spec.GCMParameterSpec(128, nonce),
        )
        return instance.doFinal(sealed)
    }
}

File: deck/base/code/native/node/runtime/cipher.ts
// AES-256-GCM over the Web Crypto API (SubtleCrypto), available on node (>=16) as globalThis.crypto.subtle. The key,
// nonce, plaintext, and ciphertext are raw bytes (the crypto currency). SubtleCrypto encrypt produces
// ciphertext || tag (16-byte GCM tag appended) and decrypt expects the same layout. Reached only through the public
// cipher API.
const cipher = (() => {
  const importKey = (key: Uint8Array): Promise<CryptoKey> =>
    globalThis.crypto.subtle.importKey(
      'raw',
      key,
      { name: 'AES-GCM' },
      false,
      ['encrypt', 'decrypt'],
    )
  return {
    encrypt: async (
      key: Uint8Array,
      nonce: Uint8Array,
      plain: Uint8Array,
    ): Promise<Uint8Array> => {
      const cryptoKey = await importKey(key)
      const sealed = await globalThis.crypto.subtle.encrypt(
        { name: 'AES-GCM', iv: nonce },
        cryptoKey,
        plain,
      )
      return new Uint8Array(sealed)
    },
    decrypt: async (
      key: Uint8Array,
      nonce: Uint8Array,
      sealed: Uint8Array,
    ): Promise<Uint8Array> => {
      const cryptoKey = await importKey(key)
      const opened = await globalThis.crypto.subtle.decrypt(
        { name: 'AES-GCM', iv: nonce },
        cryptoKey,
        sealed,
      )
      return new Uint8Array(opened)
    },
  }
})()
